### PR TITLE
Design Doc for Metapackage Proposal

### PR DESCRIPTION
**Pull request by kartiklunkad26**
## What this PR does / why we need it

This PR provides the Design Doc for the Metapackage technology which is the basis for the profile-based installation experience in #2848

## Which issue(s) this PR fixes
Provides design doc for https://github.com/vmware-tanzu/community-edition/issues/2848

### CODE REVIEW

**Comment by jorgemoralespou** [comment]
One comment I would make to this, having probably done the first metapackage and then iterated over it, and then looking at the TAP ones, is that I would make configuration more user friendly and usecase oriented, rather than package oriented and then have a proper way to translate this "user friendly" configuration to the real package configuration under the hood.

Sane defaults and enumerated configuration (configuration that packages a combined set of configuration values and that is limited and predefined) should be preferred, IMHO. An example would be to have something like `provider: [aws,docker,azure,vsphere] and based on one of those 4 values set the best sane defaults for all other packages. When you look for example to external-dns you can easily create the optimal  configuration for the user based on this instead of telling him to provide a lot of information. Just compare these 2 packages configuration [my own external-dns](https://github.com/failk8s-packages/external-dns-package#global) and [tce's bundled external-dns](https://github.com/vmware-tanzu/community-edition/tree/main/addons/packages/external-dns/0.8.0#external-dns-configuration). The latter is package oriented and imposes the user to understand how external-dns is configured. The former is user oriented, and imposes the user understand where he's using it, but not how the package is configured.

**Comment by jorgemoralespou** [comment]
Reading the proposal, will leave here some more comments:
- I would not use `Profile` as a concept for metapackages. It's very confusing and doesn't really tell what the user is doing. Not sure a better word, as to me this would be `Package` or `bundle` as a set of "canned packages that are shipped together when the XXX is selected"
- Always provide a list of packages to add or remove to a Profile, so that it can be tweaked (augmented/reduced)
- Provide a first class citizenship to `Profiles` (or whatever that end up being called) so that I can list the profiles a metapackage provides, and potentially a description. Currently this is not possible, but to me this is key, so if Carvel and or Package plugins can't be augmented, then instead of using profiles, create different metapackages per profile. (vendir can help de-duplicate code)
- As such, I would want to have a values-file being created based on the profile I will install, so that I don't see config that is not relevant to such profile.
- With regards to this section "Risks and Mitigations: UX has already been reviewed as this has been adopted by other projects using carvel technology.". The fact that the pattern has been adopted doesn't prove it's well implemented or optimal. I could argue many of the things that are needed and not easy. UX needs quite a big enhancement.
- Versioning of a metapackage with multiple profiles becomes quite complicated, as every profile follows same version scheme.

**Comment by jorgemoralespou** [comment]
All, in all, I'm in favor of doing bundles/metapackages or whatever we want to create as can be seen by this issue https://github.com/vmware-tanzu/community-edition/issues/322, but I'm personally not keen of the current implementation followed by TAP as I think can be improved. 

I would consider this proposal the place to describe what we want rather than a proposal of what we can do. TCE should be the place where we can innovate and come up with the right way of doing things later downstream.

File: docs/designs/2848-metapackage-proposal.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+This metapackage is essentially a package of packages wherein it references a Bill of Materials that includes existing packages in the repository. Being able to install, update & remove these packages in concert with each other simplifies the complexity significantly. 
+
+This is a method in practice in other projects leveraging Carvel tooling. 
+
+## Motivation
+
+The motivation for introducing this is to reduce the complexity that comes up with installing multiple carvel packages at a time. This will allow newcomers to the project rapidly establish a working technology stack.
+
+* Describe why the change is important and the benefits to users.
```
Comment (jpmcb) [inline docs/designs/2848-metapackage-proposal.md:17]: Was this part of the template? I think this line needs to be removed.

File: docs/designs/2848-metapackage-proposal.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+The proposal will describe what a metapackage is and how it will be implemented. 
+
+The proposal can be leveraged for various user scenarios. One example user scenario is described below. 
+
+**Terminology**
+
+* Metapackage - concept to bring two or more packages together. 
+* Profile - Enabling users to decide by a config value which part of the metapackage the user wants to install. 
+
+
```
Comment (joshrosso) [inline docs/designs/2848-metapackage-proposal.md:39]: nit: please remove these excess line-breaks

File: docs/designs/2848-metapackage-proposal.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+### Example Scenario
+
+The below scenario is one use-case of the metapackage technology. This is **not** the only way to use it. 
+
```
Comment (joshrosso) [inline docs/designs/2848-metapackage-proposal.md:44]: nit: please remove excess line breaks

File: docs/designs/2848-metapackage-proposal.md (hunk 1 above)
Comment (stmcginnis) [inline docs/designs/2848-metapackage-proposal.md:17]: Remove this line?

File: docs/designs/2848-metapackage-proposal.md (hunk 2 above)
Comment (jpmcb) [inline docs/designs/2848-metapackage-proposal.md:39]: Nit: too many extra lines here

Comment (jpmcb) [reply docs/designs/2848-metapackage-proposal.md:39]: The markdown linter may have also failed on these MD formatings

File: docs/designs/2848-metapackage-proposal.md (hunk 1 above)
Comment (nrb) [inline docs/designs/2848-metapackage-proposal.md:17]: This line seems left over from the template.

File: docs/designs/2848-metapackage-proposal.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+## Compatibility
+
+* tanzu cli with package plugin
+* kapp controller version
+* TCE-repo package repository
+
+## Alternatives
+
+The alternative today is to install each package individually with it's own values file and it's own upgrade planning.
```
Comment (stmcginnis) [inline docs/designs/2848-metapackage-proposal.md:149]: Super nit I only mention if you do other more important updates:

s/it's/its/

**Comment by joshrosso** [comment]
> I would consider this proposal the place to describe what we want rather than a proposal of what we can do. TCE should be the place where we can innovate and come up with the right way of doing things later downstream.

@kartiklunkad26 I second this sentiment.

So overall very much +1. However, a metapackage is essentially a package and **we can accept** these today. I'm wondering if the proposal can reflect what you're specifically looking to bring to the table as a metapackage?

Otherwise, if the question is, will TCE accept metapackage contributions? The answer is yes!

We even mention it here:

https://github.com/vmware-tanzu/community-edition/blob/main/docs/packaging/considerations/versioning.md

File: docs/designs/2848-metapackage-proposal.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+| bar 2 | bar 2 |
+|       | bar 3 |
+
+2. Installing a profile
+* List version information for the package by running:
+
+```
+tanzu package available list foo.tanzu.vmware.com 
+```
+* Create a `foo-values.yaml` file. This configuration file will have configuration required to deploy foo and the subordinate packages.
```
Comment (jpmcb) [inline docs/designs/2848-metapackage-proposal.md:78]: I don't think this should be a single yaml file: these could end up being huge and I think breaking them apart will make them easier for users to grok.

For example, lets say there was some observability metapackage that was Grafana, Prometheues, and Fluent. That could easily run 1000 lines of yaml if it was all in one.

File: docs/designs/2848-metapackage-proposal.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+### Goals
+
+* Demonstrate the ability that multiple subordinate packages can be installed using a single metapackage. 
+* Be able to provide a single values.yaml file that can be used for all the subordinate package values.yaml.
+
+### Non-Goals/Future Work
+
+* Define each profile (and the packages it will have) that will use this metapackage technology. 
+
```
Comment (jpmcb) [inline docs/designs/2848-metapackage-proposal.md:27]: Can we call out a non-goal here as _not_ being able to coordinate packages when deployed?

For example, today, _i don't believe_ it's possible via Carvel for packages to introspect one another. So if I installed a prometheus, grafana observability profile, they wouldn't be able to introspect eachother intelligently to automatically deploy dashboards into grafana based on what's on the cluster.

**Comment by nrb** [comment]
I'm a little confused by this proposal; it seems more like something for Carvel tooling...which is already supported. What is the benefit of having this in TCE?

File: docs/designs/2848-metapackage-proposal.md (hunk 6 above)
Comment (joshrosso) [reply docs/designs/2848-metapackage-proposal.md:27]: ref the idea of dependency management issue that could trend towards something like this being possible.
https://github.com/vmware-tanzu/carvel-kapp-controller/issues/371

**Comment by joshrosso** [comment]
As discussed at https://kubernetes.slack.com/archives/C02GY94A8KT/p1643999766789219, we're going to close this PR as we can accept meta packages as defined in this design doc. For us, they are largely just packages.

We'll focus on getting approval for what Tanzu Application Platform (TAP) will be brining in https://github.com/vmware-tanzu/community-edition/issues/2848.

Thanks!

**Pull request closed without merging**